### PR TITLE
add deepinv recipe

### PR DESCRIPTION
**Pull request by gschramm**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by gschramm** [comment]
@conda-forge-admin, please rerender

**Comment by tachella** [comment]
Many thanks for putting this together @gschramm!

I am happy to be listed as a maintainer 👍

**Comment by matthieutrs** [comment]
Thanks a lot @gschramm! Happy to be listed as maintainer too!

File: recipes/deepinv/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+test:
+  imports:
+    - deepinv
+  commands:
+    - pip check
+    
+  requires:
+    - pip
+    - pytest
+    - pytest deepinv/tests -vs
```
Comment (jaimergp) [inline recipes/deepinv/meta.yaml:42]: Move this line over to `commands`.  This section is only for the test dependencies.

```suggestion
```

**Comment by gschramm** [comment]
@tachella @matthieutrs seems that the builds are working, but the tests on osx take ages. right now I am running all the tests via `pytest deepinv/tests`. Do you think that all tests are needed to verify the builds?

**Comment by matthieutrs** [comment]
I would be in favour of waiting for the tests to pass but I am a bit surprised that linux takes 22 minutes while osx 5+ hours 🤔

**Comment by gschramm** [comment]
> I would be in favour of waiting for the tests to pass but I am a bit surprised that linux takes 22 minutes while osx 5+ hours 🤔

I agree. Have you ever run all tests locally on OSX?

**Comment by matthieutrs** [comment]
On osx I don't test the optional dependencies, but I test the required dependencies, maybe that's an explanation?

**Comment by gschramm** [comment]
> On osx I don't test the optional dependencies, but I test the required dependencies, maybe that's an explanation?

How do you exclude the opt. dependencies?
Is there I flag I can pass to `pytest`?

**Comment by matthieutrs** [comment]
So normally this should be automatic given our pytest format: if you `pip install -e .[test]` [as here](https://github.com/deepinv/deepinv/blob/main/.github/workflows/test.yml#L40) this should create an env without the optional dependencies ([see around these lines](https://github.com/deepinv/deepinv/blob/main/pyproject.toml#L46)) and the pytest then ignores the tests to be skipped. Now I remember that bm3d is not compatible with OSX, is this a problem? (that's why we've put it in [optional dependencies](https://github.com/deepinv/deepinv/blob/main/pyproject.toml#L63) btw)

**Comment by gschramm** [comment]
> So normally this should be automatic given our pytest format: if you `pip install -e .[test]` [as here](https://github.com/deepinv/deepinv/blob/main/.github/workflows/test.yml#L40) this should create an env without the optional dependencies ([see around these lines](https://github.com/deepinv/deepinv/blob/main/pyproject.toml#L46)) and the pytest then ignores the tests to be skipped. Now I remember that bm3d is not compatible with OSX, is this a problem? (that's why we've put it in [optional dependencies](https://github.com/deepinv/deepinv/blob/main/pyproject.toml#L63) btw)

The optional dependencies are currently not being installed - see [here](https://github.com/conda-forge/staged-recipes/blob/353deca5d3acec6103e518e5f430457457edced8/recipes/deepinv/meta.yaml#L22). So all optional tests should be skipped. 

The log of the OSX run shows that it got stuck - see [here](https://dev.azure.com/conda-forge/84710dde-1620-425b-80d0-4cf5baca359d/_apis/build/builds/900110/logs/33). But I have no real clue why.

**Comment by matthieutrs** [comment]
Thanks for the pointer! I think I may see what's going on, I'll push a fix soon

**Comment by gschramm** [comment]
> Thanks for the pointer! I think I may see what's going on, I'll push a fix soon

Ok. Should I disable the OSX build until a new version is released? (can be re-enabled as soon as the new version is out)

**Comment by matthieutrs** [comment]
I've just merged [PR #180](https://github.com/deepinv/deepinv/pull/180) that should fix it; if tests stall again, let's deactivate the OSX build until next release 👍

**Comment by gschramm** [comment]
on conda-forge we don't build from master, but from a release (tagged version from github or pypi) -  see [here](https://github.com/conda-forge/staged-recipes/blob/353deca5d3acec6103e518e5f430457457edced8/recipes/deepinv/meta.yaml#L9) so nothing will change until there is a new release

**Comment by matthieutrs** [comment]
Thanks for the clarification! Then I suggest we disable the OSX build for now.

**Comment by gschramm** [comment]
@conda-forge/help-python, ready for review!

**Comment by gschramm** [comment]
@conda-forge/help-python A few things to consider: 
- we cannot build for windows, since pytorch is only available on linux, osx
- I skipped the osx builds, because the tests took too long (and hung)
- I am bit confused by the linter saying `Non noarch packages should have python requirement without any version constraints`, `deepinv` needs python>=3.8

File: recipes/deepinv/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  version: {{ version }}
+
+source:
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/deepinv-{{ version }}.tar.gz
+  sha256: 1b44b473544e6f2998e7f80dbcb94136a3528e06dfa53b5c00c2d4c2d0bc4ec1
+
+build:
+  skip: true  # [win]
+  skip: true  # [osx]
+  skip: true  # [py<38]
```
Comment (ocefpaf) [inline recipes/deepinv/meta.yaml:15]: It is OK for noarch package to fails on Windows with a missing dependency. We prefer noarch b/c it is faster to build and the package will be ready when the missing dependency is available on Windows. 

```suggestion
  noarch: python
```

File: recipes/deepinv/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  script: {{ PYTHON }} -m pip install . -vv --no-deps --no-build-isolation
+  number: 0
+
+requirements:
+  host:
+    - python
+    - setuptools >=61.2
+    - pip
+  run:
+    - python
```
Comment (ocefpaf) [inline recipes/deepinv/meta.yaml:25]:
```suggestion
    - python >=3.8
    - setuptools >=61.2
    - pip
  run:
    - python >=3.8
```

**Comment by gschramm** [comment]
@ocefpaf 
1. I committed your changes - but I am still confused by the output of the linter, do we need the python restriction in host and run?
2. The testing on osx take too long, should we skip them?

**Comment by ocefpaf** [comment]
> 1. I committed your changes - but I am still confused by the output of the linter, do we need the python restriction in host and run?

Yes.

> 2. The testing on osx take too long, should we skip them?

No need. The noarch package will be built only on Linux after the merge.

**Pull request merged**